### ISSUE TITLE
Seqential scan in queries with `UNION ALL` (PostgreSQL)

### ISSUE THREAD

**Issue by rpiotrow** [gemini-hlsw/gsp-graphql#186]
For embedded collections Grackle generates SQL query with `UNION ALL` that is executed by PostgreSQL with sequential scans. For queries that return one entity (based on id with joins using indexes) this can be very slow (if "child" tables are big).

Below is detailed example.

Given SQL:
```sql
CREATE TABLE seq_scan_a
(
    id VARCHAR PRIMARY KEY
);

CREATE TABLE seq_scan_b
(
    id   VARCHAR PRIMARY KEY,
    a_id VARCHAR NOT NULL
);

CREATE INDEX seq_scan_b_a_id_idx ON seq_scan_b(a_id);

CREATE TABLE seq_scan_c
(
    id     VARCHAR PRIMARY KEY,
    b_id   VARCHAR NOT NULL,
    name_c VARCHAR NOT NULL
);

CREATE INDEX seq_scan_c_b_id_idx ON seq_scan_c(b_id);

CREATE TABLE seq_scan_d
(
    id     VARCHAR PRIMARY KEY,
    b_id   VARCHAR NOT NULL,
    name_d VARCHAR NOT NULL
);

CREATE INDEX seq_scan_d_b_id_idx ON seq_scan_d(b_id);

INSERT INTO seq_scan_a(id) VALUES('id_a_1');

INSERT INTO seq_scan_b(id, a_id) VALUES('id_b_1', 'id_a_1');

INSERT INTO seq_scan_c(id, b_id, name_c) VALUES('id_c_1', 'id_b_1', 'name_c');
INSERT INTO seq_scan_d(id, b_id, name_d) VALUES('id_d_1', 'id_b_1', 'name_d');
```

Grackle mapping:
```scala
import cats.implicits._
import _root_.doobie.util.meta.Meta
import _root_.doobie.util.transactor.Transactor
import cats.effect.Sync
import edu.gemini.grackle.Path.UniquePath
import edu.gemini.grackle.Predicate.{Const, Eql}
import edu.gemini.grackle.Query.{Binding, Filter, Select, Unique}
import edu.gemini.grackle._
import edu.gemini.grackle.doobie._
import edu.gemini.grackle.syntax._
import edu.gemini.grackle.QueryCompiler.SelectElaborator
import edu.gemini.grackle.Value.StringValue


trait SeqScanMapping[F[_]] extends DoobieMapping[F] {

  object aTable extends TableDef("seq_scan_a") {
    val id = col("id", Meta[String])
  }
  object bTable extends TableDef("seq_scan_b") {
    val id = col("id", Meta[String])
    val aId = col("a_id", Meta[String])
  }
  object cTable extends TableDef("seq_scan_c") {
    val id = col("id", Meta[String])
    val bId = col("b_id", Meta[String])
    val nameC = col("name_c", Meta[String])
  }
  object dTable extends TableDef("seq_scan_d") {
    val id = col("id", Meta[String])
    val bId = col("b_id", Meta[String])
    val nameD = col("name_d", Meta[String])
  }

  val schema =
    schema"""
      type Query {
        a(id: String!): A
      }
      type A {
        id: String!
        bs: [B!]!
      }
      type B {
        id: String!
        cs: [C!]!
        ds: [D!]!
      }
      type C {
        id: String!
        nameC: String!
      }
      type D {
        id: String!
        nameD: String!
      }
    """

  val QueryType = schema.ref("Query")
  val AType = schema.ref("A")
  val BType = schema.ref("B")
  val CType = schema.ref("C")
  val DType = schema.ref("D")

  val typeMappings =
    List(
      ObjectMapping(
        tpe = QueryType,
        fieldMappings =
          List(
            SqlRoot("a")
          )
      ),
      ObjectMapping(
        tpe = AType,
        fieldMappings =
          List(
            SqlField("id", aTable.id, key = true, hidden = true),
            SqlObject("bs", Join(aTable.id, bTable.aId))
          )
      ),
      ObjectMapping(
        tpe = BType,
        fieldMappings =
          List(
            SqlField("id", bTable.id, key = true, hidden = true),
            SqlObject("cs", Join(bTable.id, cTable.bId)),
            SqlObject("ds", Join(bTable.id, dTable.bId))
          )
      ),
      ObjectMapping(
        tpe = CType,
        fieldMappings =
          List(
            SqlField("id", cTable.id, key = true, hidden = true),
            SqlField("nameC", cTable.nameC)
          )
      ),
      ObjectMapping(
        tpe = DType,
        fieldMappings =
          List(
            SqlField("id", dTable.id, key = true, hidden = true),
            SqlField("nameD", dTable.nameD)
          )
      )
    )

  override val selectElaborator: SelectElaborator = new SelectElaborator(
    Map(
      QueryType -> {
        case Select("a", List(Binding("id", StringValue(id))), child) =>
          Select("a", Nil, Unique(Filter(Eql(UniquePath[String](List("id")), Const(id)), child))).rightIor
        case other => other.rightIor
      }
    )
  )

}

object SeqScanMapping extends LoggedDoobieMappingCompanion {

  def mkMapping[F[_]: Sync](transactor: Transactor[F], monitor: DoobieMonitor[F]): Mapping[F] =
    new DoobieMapping[F](transactor, monitor) with SeqScanMapping[F]

}
```

GraphQL Query:
```
query {
  a(id: "id_a_1") {
    bs {
      cs { nameC }
      ds { nameD }
    }
  }
}
```
generates SQL:
```sql
SELECT seq_scan_a.id,
       seq_scan_b_alias_0.name_d_of_seq_scan_d,
       seq_scan_b_alias_0.id_of_seq_scan_d,
       seq_scan_b_alias_0.id_of_seq_scan_b,
       seq_scan_b_alias_0.a_id_of_seq_scan_b,
       seq_scan_b_alias_0.name_c_of_seq_scan_c,
       seq_scan_b_alias_0.id_of_seq_scan_c
FROM
  (SELECT seq_scan_a.id
   FROM seq_scan_a
   INNER JOIN
     (SELECT DISTINCT seq_scan_a.id
      FROM seq_scan_a
      WHERE (seq_scan_a.id IS NOT NULL)
        AND (seq_scan_a.id = 'id_a_1') ) AS pred_0 ON (pred_0.id = seq_scan_a.id)) AS seq_scan_a
LEFT JOIN (
             (SELECT seq_scan_d.name_d AS name_d_of_seq_scan_d,
                     seq_scan_d.id AS id_of_seq_scan_d,
                     seq_scan_b.id AS id_of_seq_scan_b,
                     seq_scan_b.a_id AS a_id_of_seq_scan_b,
                     (NULL :: VARCHAR) AS name_c_of_seq_scan_c,
                     (NULL :: VARCHAR) AS id_of_seq_scan_c
              FROM seq_scan_b
              LEFT JOIN seq_scan_d ON (seq_scan_d.b_id = seq_scan_b.id))
           UNION ALL
             (SELECT (NULL :: VARCHAR) AS name_d_of_seq_scan_d,
                     (NULL :: VARCHAR) AS id_of_seq_scan_d,
                     seq_scan_b.id AS id_of_seq_scan_b,
                     seq_scan_b.a_id AS a_id_of_seq_scan_b,
                     seq_scan_c.name_c AS name_c_of_seq_scan_c,
                     seq_scan_c.id AS id_of_seq_scan_c
              FROM seq_scan_b
              LEFT JOIN seq_scan_c ON (seq_scan_c.b_id = seq_scan_b.id))) AS seq_scan_b_alias_0 ON (seq_scan_b_alias_0.a_id_of_seq_scan_b = seq_scan_a.id);
```

Qeury execution plan (PostgreSQL):
```
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Right Join  (cost=46.18..145.42 rows=1 width=224) (actual time=0.088..0.135 rows=2 loops=1)
   Hash Cond: ((seq_scan_b.a_id)::text = (seq_scan_a.id)::text)
   ->  Append  (cost=29.80..122.43 rows=1760 width=192) (actual time=0.029..0.065 rows=2 loops=1)
         ->  Hash Right Join  (cost=29.80..48.02 rows=880 width=192) (actual time=0.028..0.038 rows=1 loops=1)
               Hash Cond: ((seq_scan_d.b_id)::text = (seq_scan_b.id)::text)
               ->  Seq Scan on seq_scan_d  (cost=0.00..16.50 rows=650 width=96) (actual time=0.003..0.004 rows=1 loops=1)
               ->  Hash  (cost=18.80..18.80 rows=880 width=64) (actual time=0.012..0.013 rows=1 loops=1)
                     Buckets: 1024  Batches: 1  Memory Usage: 9kB
                     ->  Seq Scan on seq_scan_b  (cost=0.00..18.80 rows=880 width=64) (actual time=0.008..0.009 rows=1 loops=1)
         ->  Hash Right Join  (cost=29.80..48.02 rows=880 width=192) (actual time=0.015..0.023 rows=1 loops=1)
               Hash Cond: ((seq_scan_c.b_id)::text = (seq_scan_b_1.id)::text)
               ->  Seq Scan on seq_scan_c  (cost=0.00..16.50 rows=650 width=96) (actual time=0.003..0.004 rows=1 loops=1)
               ->  Hash  (cost=18.80..18.80 rows=880 width=64) (actual time=0.004..0.005 rows=1 loops=1)
                     Buckets: 1024  Batches: 1  Memory Usage: 9kB
                     ->  Seq Scan on seq_scan_b seq_scan_b_1  (cost=0.00..18.80 rows=880 width=64) (actual time=0.002..0.003 rows=1 loops=1)
   ->  Hash  (cost=16.36..16.36 rows=1 width=32) (actual time=0.047..0.049 rows=1 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 9kB
         ->  Nested Loop  (cost=0.30..16.36 rows=1 width=32) (actual time=0.040..0.045 rows=1 loops=1)
               ->  Unique  (cost=0.15..8.17 rows=1 width=32) (actual time=0.028..0.032 rows=1 loops=1)
                     ->  Index Only Scan using seq_scan_a_pkey on seq_scan_a seq_scan_a_1  (cost=0.15..8.17 rows=1 width=32) (actual time=0.026..0.028 rows=1 loops=1)
                           Index Cond: ((id IS NOT NULL) AND (id = 'id_a_1'::text))
                           Heap Fetches: 1
               ->  Index Only Scan using seq_scan_a_pkey on seq_scan_a  (cost=0.15..8.17 rows=1 width=32) (actual time=0.005..0.005 rows=1 loops=1)
                     Index Cond: (id = (seq_scan_a_1.id)::text)
                     Heap Fetches: 1
 Planning Time: 0.840 ms
 Execution Time: 0.246 ms
(27 rows)
```

**Comment by milessabin** [comment]
Current WIP generates the following SQL,

```sql
(
  SELECT
    seq_scan_d.name_d,
    seq_scan_d.id AS id_alias_0,
    seq_scan_b.id AS id_alias_1,
    seq_scan_a.id,
    (NULL :: VARCHAR) AS name_c,
    (NULL :: VARCHAR) AS id_alias_2
  FROM
    (
      SELECT
        seq_scan_a.id
      FROM
        seq_scan_a
        INNER JOIN (
          SELECT
            DISTINCT seq_scan_a.id
          FROM
            seq_scan_a
          WHERE
            (seq_scan_a.id IS NOT NULL)
            AND (seq_scan_a.id = 'id_a_1')
        ) AS pred ON (pred.id = seq_scan_a.id)
    ) AS seq_scan_a
    LEFT JOIN seq_scan_b ON (seq_scan_b.a_id = seq_scan_a.id)
    LEFT JOIN seq_scan_d ON (seq_scan_d.b_id = seq_scan_b.id)
)
UNION ALL
(
  SELECT
    (NULL :: VARCHAR) AS name_d,
    (NULL :: VARCHAR) AS id_alias_0,
    seq_scan_b.id AS id_alias_1,
    seq_scan_a.id,
    seq_scan_c.name_c,
    seq_scan_c.id AS id_alias_2
  FROM
    (
      SELECT
        seq_scan_a.id
      FROM
        seq_scan_a
        INNER JOIN (
          SELECT
            DISTINCT seq_scan_a.id
          FROM
            seq_scan_a
          WHERE
            (seq_scan_a.id IS NOT NULL)
            AND (seq_scan_a.id = 'id_a_1')
        ) AS pred ON (pred.id = seq_scan_a.id)
    ) AS seq_scan_a
    LEFT JOIN seq_scan_b ON (seq_scan_b.a_id = seq_scan_a.id)
    LEFT JOIN seq_scan_c ON (seq_scan_c.b_id = seq_scan_b.id)
)
```
You'll notice that the predicate has been pushed down into both branches of the `UNION ALL`. The corresponding `EXPLAIN ANALYZE` shows that the `Seq Scan` has been eliminated,
```
                                                                                 QUERY PLAN                                                                                  
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Append  (cost=0.60..33.92 rows=2 width=192) (actual time=0.042..0.066 rows=2 loops=1)
   ->  Nested Loop Left Join  (cost=0.60..16.95 rows=1 width=192) (actual time=0.042..0.043 rows=1 loops=1)
         ->  Nested Loop Left Join  (cost=0.45..16.66 rows=1 width=64) (actual time=0.029..0.030 rows=1 loops=1)
               ->  Nested Loop  (cost=0.30..16.36 rows=1 width=32) (actual time=0.024..0.025 rows=1 loops=1)
                     ->  Unique  (cost=0.15..8.17 rows=1 width=32) (actual time=0.017..0.017 rows=1 loops=1)
                           ->  Index Only Scan using seq_scan_a_pkey on seq_scan_a seq_scan_a_1  (cost=0.15..8.17 rows=1 width=32) (actual time=0.016..0.016 rows=1 loops=1)
                                 Index Cond: ((id IS NOT NULL) AND (id = 'id_a_1'::text))
                                 Heap Fetches: 1
                     ->  Index Only Scan using seq_scan_a_pkey on seq_scan_a  (cost=0.15..8.17 rows=1 width=32) (actual time=0.005..0.005 rows=1 loops=1)
                           Index Cond: (id = (seq_scan_a_1.id)::text)
                           Heap Fetches: 1
               ->  Index Scan using seq_scan_b_a_id_idx on seq_scan_b  (cost=0.15..0.26 rows=4 width=64) (actual time=0.004..0.004 rows=1 loops=1)
                     Index Cond: ((a_id)::text = (seq_scan_a.id)::text)
         ->  Index Scan using seq_scan_d_b_id_idx on seq_scan_d  (cost=0.15..0.26 rows=3 width=96) (actual time=0.011..0.012 rows=1 loops=1)
               Index Cond: ((b_id)::text = (seq_scan_b.id)::text)
   ->  Nested Loop Left Join  (cost=0.60..16.95 rows=1 width=192) (actual time=0.022..0.023 rows=1 loops=1)
         ->  Nested Loop Left Join  (cost=0.45..16.66 rows=1 width=64) (actual time=0.015..0.015 rows=1 loops=1)
               ->  Nested Loop  (cost=0.30..16.36 rows=1 width=32) (actual time=0.010..0.010 rows=1 loops=1)
                     ->  Unique  (cost=0.15..8.17 rows=1 width=32) (actual time=0.005..0.005 rows=1 loops=1)
                           ->  Index Only Scan using seq_scan_a_pkey on seq_scan_a seq_scan_a_3  (cost=0.15..8.17 rows=1 width=32) (actual time=0.005..0.005 rows=1 loops=1)
                                 Index Cond: ((id IS NOT NULL) AND (id = 'id_a_1'::text))
                                 Heap Fetches: 1
                     ->  Index Only Scan using seq_scan_a_pkey on seq_scan_a seq_scan_a_2  (cost=0.15..8.17 rows=1 width=32) (actual time=0.004..0.004 rows=1 loops=1)
                           Index Cond: (id = (seq_scan_a_3.id)::text)
                           Heap Fetches: 1
               ->  Index Scan using seq_scan_b_a_id_idx on seq_scan_b seq_scan_b_1  (cost=0.15..0.26 rows=4 width=64) (actual time=0.004..0.004 rows=1 loops=1)
                     Index Cond: ((a_id)::text = (seq_scan_a_2.id)::text)
         ->  Index Scan using seq_scan_c_b_id_idx on seq_scan_c  (cost=0.15..0.26 rows=3 width=96) (actual time=0.006..0.007 rows=1 loops=1)
               Index Cond: ((b_id)::text = (seq_scan_b_1.id)::text)
 Planning Time: 0.783 ms
 Execution Time: 0.162 ms
```

**Comment by milessabin** [comment]
Fixed in #188.
